File: Sources/ExcalidrawModel/Restore.swift
import Foundation

/// Load-time normalisation, the Swift counterpart to `packages/excalidraw/data/
/// restore.ts`. This is the single entry point all loaded files should pass
/// through. Phase 1 implements the highest-value, safe normalisations; richer
/// migrations (legacy binding v1 → v2, container repair, oversized-arrow
/// clamping) land incrementally and each gets a regression fixture.
public enum Restore {
    public static func restore(_ file: ExcalidrawFile) -> ExcalidrawFile {
        var file = file
        file.type = ExcalidrawSchema.fileType
        if file.version < ExcalidrawSchema.schemaVersion {
            file.version = ExcalidrawSchema.schemaVersion
        }
        file.elements = restoreElements(file.elements)
        return file
    }

    static func restoreElements(_ elements: [ExcalidrawElement]) -> [ExcalidrawElement] {
        var result = elements
        assignMissingIndices(&result)
        return result
    }

    /// Ensure every element carries a fractional `index`. Existing indices are
    /// preserved; gaps are filled with monotonically increasing keys derived
    /// from array order.
    ///
    /// NOTE: this is a simplified, lexicographically-monotone generator, not the
    /// full rocicorp fractional-indexing algorithm. It is sufficient for stable
    /// ordering within a document; the complete algorithm (needed for conflict-
    /// free concurrent inserts) is a later increment.
    static func assignMissingIndices(_ elements: inout [ExcalidrawElement]) {
        guard elements.contains(where: { $0.base.index == nil }) else { return }
        for i in elements.indices where elements[i].base.index == nil {
            elements[i].base.index = FractionalIndex.key(forPosition: i)
        }
    }
}

/// Minimal fractional-index key support. Fixed-width keys sort lexicographically
/// in the same order as their numeric position.
enum FractionalIndex {
    static func key(forPosition position: Int) -> String {
        // 7 base-36 digits keep keys ordered for millions of elements.
        let digits = String(position, radix: 36)
        let padded = String(repeating: "0", count: max(0, 7 - digits.count)) + digits
        return "a" + padded
    }
}

File: web/packages/excalidraw-svelte/src/model/restore.ts
import type { ExcalidrawElement } from "./element.js";
import { ExcalidrawSchema } from "./enums.js";
import type { ExcalidrawFile } from "./file.js";

/**
 * Minimal fractional-index key support. Fixed-width base-36 keys sort
 * lexicographically in the same order as their numeric position. (parity:
 * Restore.swift / FractionalIndex)
 */
export function fractionalIndexKey(position: number): string {
  const digits = position.toString(36);
  const padded = "0".repeat(Math.max(0, 7 - digits.length)) + digits;
  return `a${padded}`;
}

/** Ensure every element carries a fractional `index`; existing keys are kept. */
function assignMissingIndices(elements: ExcalidrawElement[]): ExcalidrawElement[] {
  if (!elements.some((el) => el.index === null)) return elements;
  return elements.map((el, i) =>
    el.index === null ? { ...el, index: fractionalIndexKey(i) } : el,
  );
}

function restoreElements(elements: ExcalidrawElement[]): ExcalidrawElement[] {
  return assignMissingIndices(elements);
}

/**
 * Load-time normalisation, the counterpart to `restore.ts`. The single entry
 * point all loaded files pass through: canonicalise the envelope and backfill
 * missing element indices. (parity: Restore.swift)
 */
export function restore(file: ExcalidrawFile): ExcalidrawFile {
  return {
    ...file,
    type: ExcalidrawSchema.fileType,
    version:
      file.version < ExcalidrawSchema.schemaVersion ? ExcalidrawSchema.schemaVersion : file.version,
    elements: restoreElements(file.elements),
  };
}
